mapbox/libandroid-telemetry/src/main/java/com/mapbox/services/android/telemetry/navigation/MapboxNavigationEvent.java

```diff
@@ -0,0 +1,169 @@
+  public static final String KEY_START_TIMESTAMP = "startTimestamp";
+  public static final String KEY_DISTANCE_COMPLETED = "distanceCompleted";
+  public static final String KEY_DISTANCE_REMAINING = "distanceRemaining";
+  public static final String KEY_DURATION_REMAINING = "durationRemaining";
+  public static final String KEY_SECONDS_SINCE_LAST_REROUTE = "secondsSinceLastReroute";
+
+  /**
+   * Navigation turnstile.
+   */
+  public static Hashtable<String, Object> buildTurnstileEvent(String sdKIdentifier, String sdkVersion) {
+    Hashtable<String, Object> event = new Hashtable<>();
+    event.put(KEY_EVENT, TYPE_TURNSTILE);
+    event.put(KEY_PLATFORM, TelemetryConstants.PLATFORM);
+    event.put(KEY_OPERATING_SYSTEM, TelemetryConstants.OPERATING_SYSTEM);
+    event.put(KEY_SDK_IDENTIFIER, sdKIdentifier);
+    event.put(KEY_SDK_VERSION, sdkVersion);
+    return event;
+  }
+
+  /**
+   * User started a route.
+   */
+  public static Hashtable<String, Object> buildDepartEvent(
+      String sdKIdentifier, String sdkVersion, String sessionIdentifier, double lat, double lng,
+      String geometry, String profile, int estimatedDistance, int estimatedDuration,
+      int rerouteCount) {
+    Hashtable<String, Object> event = getMetadata(sdKIdentifier, sdkVersion, sessionIdentifier,
+        lat, lng, geometry, profile, estimatedDistance, estimatedDuration, rerouteCount);
+    event.put(KEY_EVENT, TYPE_DEPART);
+    return event;
+  }
+
+  /**
+   * User feedback/reroute event.
+   */
+  public static Hashtable<String, Object> buildFeedbackEvent(
+      String sdKIdentifier, String sdkVersion, String sessionIdentifier, double lat, double lng,
+      String geometry, String profile, int estimatedDistance, int estimatedDuration,
+      int rerouteCount, Date startTimestamp, String feedbackType,
+      Location[] locationsBefore, Location[] locationsAfter, int distanceCompleted,
+      int distanceRemaining, int durationRemaining, int newDistanceRemaining,
+      int newDurationRemaining, int secondsSinceLastReroute) {
+    Hashtable<String, Object> event = getMetadata(sdKIdentifier, sdkVersion, sessionIdentifier,
+        lat, lng, geometry, profile, estimatedDistance, estimatedDuration, rerouteCount);
+    event.put(KEY_EVENT, TYPE_FEEDBACK);
+    event.put(KEY_START_TIMESTAMP, TelemetryUtils.generateCreateDateFormatted(startTimestamp));
+    event.put(KEY_FEEDBACK_TYPE, feedbackType);
+    event.put(KEY_LOCATIONS_BEFORE, locationsBefore);
+    event.put(KEY_LOCATIONS_AFTER, locationsAfter);
+    event.put(KEY_DISTANCE_COMPLETED, distanceCompleted);
+    event.put(KEY_DISTANCE_REMAINING, distanceRemaining);
+    event.put(KEY_DURATION_REMAINING, durationRemaining);
+    event.put(KEY_NEW_DISTANCE_REMAINING, newDistanceRemaining);
+    event.put(KEY_NEW_DURATION_REMAINING, newDurationRemaining);
+    event.put(KEY_SECONDS_SINCE_LAST_REROUTE, secondsSinceLastReroute);
```

Comment:
`newDistanceRemaining`, `newDurationRemaining`, and `secondsSinceLastReroute` are only required when the `feedbackType=="reroute"`, they should be empty or null otherwise.